File: vortexdbg-android/src/main/java/com/vortexdbg/linux/android/dvm/ArmVarArg64.java
package com.vortexdbg.linux.android.dvm;

import com.vortexdbg.Emulator;
import com.vortexdbg.pointer.VortexdbgPointer;
import unicorn.Arm64Const;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;

class ArmVarArg64 extends ArmVarArg {

    ArmVarArg64(Emulator<?> emulator, BaseVM vm, DvmMethod method) {
        super(emulator, vm, method);

        int offset = 0;
        int floatOff = 0;
        for (Shorty shorty : shorties) {
            switch (shorty.getType()) {
                case 'L':
                case 'B':
                case 'C':
                case 'I':
                case 'S':
                case 'Z': {
                    args.add(getInt(offset++));
                    break;
                }
                case 'D': {
                    args.add(getVectorArg(floatOff++));
                    break;
                }
                case 'F': {
                    args.add((float) getVectorArg(floatOff++));
                    break;
                }
                case 'J': {
                    VortexdbgPointer ptr = getArg(offset++);
                    args.add(ptr == null ? 0L : ptr.peer);
                    break;
                }
                default:
                    throw new IllegalStateException("c=" + shorty.getType());
            }
        }
    }

    private double getVectorArg(int index) {
        ByteBuffer buffer = ByteBuffer.allocate(16);
        buffer.order(ByteOrder.LITTLE_ENDIAN);
        buffer.put(emulator.getBackend().reg_read_vector(Arm64Const.UC_ARM64_REG_Q0 + index));
        buffer.flip();
        return buffer.getDouble();
    }
}

File: vortexdbg-android/src/main/kotlin/com/vortexdbg/linux/android/dvm/ArmVarArg64.kt
package com.vortexdbg.linux.android.dvm

import com.vortexdbg.Emulator
import com.vortexdbg.pointer.VortexdbgPointer
import unicorn.Arm64Const

import java.nio.ByteBuffer
import java.nio.ByteOrder

internal class ArmVarArg64(emulator: Emulator<*>, vm: BaseVM, method: DvmMethod) : ArmVarArg(emulator, vm, method) {

    init {
        var offset = 0
        var floatOff = 0
        for (shorty in shorties) {
            when (shorty.getType()) {
                'L', 'B', 'C', 'I', 'S', 'Z' -> {
                    args.add(getInt(offset++))
                }
                'D' -> {
                    args.add(getVectorArg(floatOff++))
                }
                'F' -> {
                    args.add(getVectorArg(floatOff++).toFloat())
                }
                'J' -> {
                    val ptr = getArg(offset++)
                    args.add(if (ptr == null) 0L else ptr.peer)
                }
                else -> throw IllegalStateException("c=" + shorty.getType())
            }
        }
    }

    private fun getVectorArg(index: Int): Double {
        val buffer = ByteBuffer.allocate(16)
        buffer.order(ByteOrder.LITTLE_ENDIAN)
        buffer.put(emulator.getBackend().reg_read_vector(Arm64Const.UC_ARM64_REG_Q0 + index)!!)
        buffer.flip()
        return buffer.getDouble()
    }
}
